### PR TITLE
KEYCLOAK-16918 Set custom user attribute to Name ID Format for a SAML client

### PR DESCRIPTION
**Pull request by i7a7467**
https://issues.redhat.com/projects/KEYCLOAK/issues/KEYCLOAK-16918
https://groups.google.com/g/keycloak-dev/c/4ExUdT9rc2g

### SAMLNameIdMapper interface
I created SAMLNameIdMapper interface so that we can add Script mapper etc. in the future.

### SAML Metadata
I do not change NameID Format  in SAML 2.0 Identity Provider Metadata.
I think it is unclear whether NameID will be the expected value.

Docs PR : https://github.com/keycloak/keycloak-documentation/pull/1149

### CODE REVIEW

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you very much for PR @i7a7467. It looks good to me, I just found one minor issue in one of the test cases. Otherwise, it looks good to me.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/NameIdMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                ProtocolMapperUtils.USER_ATTRIBUTE,"email"))
+                .update();
+        testExpectedNameId(bburkeUser.getEmail());
+    }
+
+    @Test
+    public void testNameIdMapperNotFound() {
+        pmu.add(createSamlProtocolMapper(UserAttributeNameIdMapper.PROVIDER_ID,
+                NameIdMapperHelper.MAPPER_NAMEID_FORMAT, JBossSAMLURIConstants.NAMEID_FORMAT_EMAIL.get(),
+                ProtocolMapperUtils.USER_ATTRIBUTE, "username"))
```
[the remark this answers is not in the document]
Comment (i7a7467) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/NameIdMapperTest.java:62]: Sorry. Yes, you're right. It's better not to use username.

**Review by mhajas — DISMISSED** [review]
Thank you for the fix @i7a7467! The PR LGTM now. @hmlnarik Can you have a look and possibly merge?

**Comment by mhajas** [comment]
I forgot about one more thing. @i7a7467 Can you squash the commits into one please?

**Comment by i7a7467** [comment]
Thanks. @mhajas 
I squashed the commit.

**Review by mhajas — DISMISSED** [review]

**Review by hmlnarik — DISMISSED** [review]
Approving based on @mhajas's approval

**Comment by hmlnarik** [comment]
@i7a7467 Since there's a conflict, could you please rebase to current master?

**Review by hmlnarik — APPROVED** [review]
Thank you @i7a7467. Reapproving

**Pull request merged**